### PR TITLE
Rule engine

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/Engine.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      e.printStackTrace();
+    }
+  }
+
+  public Engine(List<CommonRule> rules) {
+    this.rules = rules;
+    this.engineContext = new EngineContext();
+    for (CommonRule rule : rules) {
+      rule.setContext(engineContext);
+    }
```
Comment (ghost) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/Engine.java]: Not important, but any reason to use loops instead of `.forEach`?

**Review by ghost — APPROVED** [review]

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/Engine.java (hunk 1 above)
Comment (saberduck) [reply common-rule-engine/src/main/java/org/sonar/commonruleengine/Engine.java]: No particular reason, fixed

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/Engine.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  public List<Issue> scan(UastNode uast) {
```
Comment (alban-auzeill) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/Engine.java]: we should clean "engineContext" from the previous scan, to not append issues to the previous result.

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/Issue.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    return rule;
+  }
+
+  public UastNode getNode() {
+    return node;
+  }
+
+  @Override
+  public String toString() {
+    return String.format("%s %s: %s", node, rule.getClass().getSimpleName(), message);
```
Comment (ghost) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/Issue.java:32]: I would display the position here, by explicitly declaring the `toString()` method from `Node`

File: common-rule-engine/src/main/java/org/sonar/uast/Uast.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package org.sonar.uast;
+
+import com.google.gson.Gson;
+import java.io.Reader;
+
+public class Uast {
```
Comment (ghost) [inline common-rule-engine/src/main/java/org/sonar/uast/Uast.java]: could be final

File: build.gradle [hunk 5]
```diff
@@ -0,0 +1,6 @@
+/*
+ * This file was generated by the Gradle 'init' task.
+ *
+ * This is a general purpose Gradle build.
+ * Learn how to create Gradle builds at https://guides.gradle.org/creating-new-gradle-builds/
+ */
```
Comment (alban-auzeill) [inline build.gradle]: This file could be empty (even missing)

File: common-rule-engine/build.gradle [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+repositories {
+    mavenCentral()
+}
+
+apply plugin: 'java-library'
+apply plugin: 'org.junit.platform.gradle.plugin'
+
+ext.junitJupiterVersion  = '5.1.0'
+ext.gsonVersion  = '2.8.2'
+ext.guavaVersion  = '24.0-jre'
```
Comment (alban-auzeill) [inline common-rule-engine/build.gradle]: guava is not used, and not useful IMO

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/Engine.java (hunk 2 above)
Comment (saberduck) [reply common-rule-engine/src/main/java/org/sonar/commonruleengine/Engine.java]: in fact we can create new context for each scan, fixed

File: build.gradle (hunk 5 above)
Comment (saberduck) [reply build.gradle]: indeed, deleted

File: common-rule-engine/build.gradle (hunk 6 above)
Comment (saberduck) [reply common-rule-engine/build.gradle]: removed

**Pull request merged**